File: samples/CameraAccess/CameraAccess/OpenClaw/OpenClawEventClient.swift
import Foundation

class OpenClawEventClient {
  var onNotification: ((String) -> Void)?

  private var webSocketTask: URLSessionWebSocketTask?
  private var session: URLSession?
  private var isConnected = false
  private var shouldReconnect = false
  private var reconnectDelay: TimeInterval = 2
  private let maxReconnectDelay: TimeInterval = 30

  // Pending RPC responses and chat results
  private var pendingResponses: [String: ([String: Any]) -> Void] = [:]
  private var pendingChatResults: [String: (String?) -> Void] = [:]

  func connect() {
    guard GeminiConfig.isOpenClawConfigured else {
      NSLog("[OpenClawWS] Not configured, skipping")
      return
    }

    shouldReconnect = true
    reconnectDelay = 2
    establishConnection()
  }

  func disconnect() {
    shouldReconnect = false
    isConnected = false
    // Cancel all pending callbacks so they don't fire after session stops
    pendingResponses.removeAll()
    pendingChatResults.removeAll()
    webSocketTask?.cancel(with: .normalClosure, reason: nil)
    webSocketTask = nil
    session?.invalidateAndCancel()
    session = nil
    NSLog("[OpenClawWS] Disconnected")
  }

  // MARK: - Private

  private func establishConnection() {
    let host = GeminiConfig.openClawHost
      .replacingOccurrences(of: "http://", with: "")
      .replacingOccurrences(of: "https://", with: "")
    let port = GeminiConfig.openClawPort
    guard let url = URL(string: "ws://\(host):\(port)") else {
      NSLog("[OpenClawWS] Invalid URL")
      return
    }

    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    session = URLSession(configuration: config)
    webSocketTask = session?.webSocketTask(with: url)
    webSocketTask?.resume()

    NSLog("[OpenClawWS] Connecting to %@", url.absoluteString)
    startReceiving()
  }

  private func startReceiving() {
    webSocketTask?.receive { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(let message):
        switch message {
        case .string(let text):
          self.handleMessage(text)
        case .data(let data):
          if let text = String(data: data, encoding: .utf8) {
            self.handleMessage(text)
          }
        @unknown default:
          break
        }
        self.startReceiving()
      case .failure(let error):
        NSLog("[OpenClawWS] Receive error: %@", error.localizedDescription)
        self.isConnected = false
        self.scheduleReconnect()
      }
    }
  }

  private func handleMessage(_ text: String) {
    guard let data = text.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let type = json["type"] as? String else { return }

    if type == "event" {
      handleEvent(json)
    } else if type == "res" {
      let id = json["id"] as? String ?? ""
      if let callback = pendingResponses.removeValue(forKey: id) {
        callback(json)
      } else {
        let ok = json["ok"] as? Bool ?? false
        if ok {
          NSLog("[OpenClawWS] Connected and authenticated")
          isConnected = true
          reconnectDelay = 2
        } else {
          let error = json["error"] as? [String: Any]
          let msg = error?["message"] as? String ?? "unknown"
          NSLog("[OpenClawWS] Connect failed: %@", msg)
        }
      }
    }
  }

  private func handleEvent(_ json: [String: Any]) {
    guard let event = json["event"] as? String else { return }
    let payload = json["payload"] as? [String: Any] ?? [:]

    switch event {
    case "connect.challenge":
      sendConnectHandshake()

    case "heartbeat":
      handleHeartbeatEvent(payload)

    case "cron":
      handleCronEvent(payload)

    case "chat":
      handleChatEvent(payload)

    default:
      break
    }
  }

  private func sendConnectHandshake() {
    let connectMsg: [String: Any] = [
      "type": "req",
      "id": UUID().uuidString,
      "method": "connect",
      "params": [
        "minProtocol": 1,
        "maxProtocol": 1,
        "client": [
          "id": "gateway-client",
          "displayName": "VisionClaw Glass",
          "version": "1.0",
          "platform": "ios",
          "mode": "backend"
        ],
        "auth": [
          "token": GeminiConfig.openClawGatewayToken
        ]
      ] as [String: Any]
    ]

    guard let data = try? JSONSerialization.data(withJSONObject: connectMsg),
          let string = String(data: data, encoding: .utf8) else { return }
    webSocketTask?.send(.string(string)) { error in
      if let error {
        NSLog("[OpenClawWS] Handshake send error: %@", error.localizedDescription)
      }
    }
  }

  private func handleHeartbeatEvent(_ payload: [String: Any]) {
    let status = payload["status"] as? String ?? ""
    // Only notify if there's actual content (not empty/silent heartbeats)
    guard status == "sent", let preview = payload["preview"] as? String, !preview.isEmpty else {
      return
    }

    let silent = payload["silent"] as? Bool ?? false
    guard !silent else { return }

    NSLog("[OpenClawWS] Heartbeat notification: %@", String(preview.prefix(100)))
    onNotification?("[Notification from your assistant] \(preview)")
  }

  private func handleCronEvent(_ payload: [String: Any]) {
    let action = payload["action"] as? String ?? ""
    guard action == "finished" else { return }

    let summary = payload["summary"] as? String
      ?? payload["result"] as? String
      ?? ""
    guard !summary.isEmpty else { return }

    NSLog("[OpenClawWS] Cron notification: %@", String(summary.prefix(100)))
    onNotification?("[Scheduled update] \(summary)")
  }

  private func handleChatEvent(_ payload: [String: Any]) {
    let state = payload["state"] as? String ?? ""
    let runId = payload["runId"] as? String ?? ""
    guard !runId.isEmpty else { return }

    if state == "final" {
      if let callback = pendingChatResults.removeValue(forKey: runId) {
        let message = payload["message"] as? [String: Any]
        let content = message?["content"]
        let replyText: String?
        if let text = content as? String {
          replyText = text
        } else if let parts = content as? [[String: Any]] {
          replyText = parts.compactMap { ($0["type"] as? String == "text") ? $0["text"] as? String : nil }.joined(separator: "\n")
        } else {
          replyText = nil
        }
        NSLog("[OpenClawWS] chat final for %@: %@", runId, String((replyText ?? "nil").prefix(200)))
        callback(replyText ?? "Agent completed but returned no text.")
      }
    } else if state == "error" {
      if let callback = pendingChatResults.removeValue(forKey: runId) {
        let errorMsg = payload["errorMessage"] as? String ?? "Agent error"
        NSLog("[OpenClawWS] chat error for %@: %@", runId, errorMsg)
        callback(nil)
      }
    }
  }

  /// Send a chat message with optional image attachment via WebSocket chat.send RPC.
  /// This is the only way to reliably pass images to the OpenClaw agent.
  func sendChatMessage(
    sessionKey: String,
    message: String,
    imageBase64: String? = nil,
    completion: @escaping (String?) -> Void
  ) {
    guard isConnected, webSocketTask != nil else {
      NSLog("[OpenClawWS] Cannot send chat.send: not connected")
      completion(nil)
      return
    }

    let reqId = UUID().uuidString
    var params: [String: Any] = [
      "sessionKey": sessionKey,
      "message": message,
      "idempotencyKey": reqId
    ]

    if let imageBase64 {
      params["attachments"] = [[
        "mimeType": "image/jpeg",
        "fileName": "camera_frame.jpg",
        "content": imageBase64
      ]]
    }

    let request: [String: Any] = [
      "type": "req",
      "id": reqId,
      "method": "chat.send",
      "params": params
    ]

    // Register RPC ack callback — then wait for chat event
    pendingResponses[reqId] = { [weak self] response in
      let ok = response["ok"] as? Bool ?? false
      if ok {
        NSLog("[OpenClawWS] chat.send accepted, waiting for agent reply (runId=%@)", reqId)
        self?.pendingChatResults[reqId] = completion
      } else {
        let error = response["error"] as? [String: Any]
        let msg = error?["message"] as? String ?? "unknown"
        NSLog("[OpenClawWS] chat.send rejected: %@", msg)
        completion(nil)
      }
    }

    guard let data = try? JSONSerialization.data(withJSONObject: request),
          let string = String(data: data, encoding: .utf8) else {
      pendingResponses.removeValue(forKey: reqId)
      completion(nil)
      return
    }

    webSocketTask?.send(.string(string)) { [weak self] error in
      if let error {
        NSLog("[OpenClawWS] chat.send send error: %@", error.localizedDescription)
        self?.pendingResponses.removeValue(forKey: reqId)
        completion(nil)
      } else {
        NSLog("[OpenClawWS] chat.send sent (id=%@, hasImage=%@)", reqId, imageBase64 != nil ? "true" : "false")
      }
    }
  }

  private func scheduleReconnect() {
    guard shouldReconnect else { return }
    NSLog("[OpenClawWS] Reconnecting in %.0fs", reconnectDelay)
    DispatchQueue.main.asyncAfter(deadline: .now() + reconnectDelay) { [weak self] in
      guard let self, self.shouldReconnect else { return }
      self.reconnectDelay = min(self.reconnectDelay * 2, self.maxReconnectDelay)
      self.establishConnection()
    }
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/openclaw/OpenClawEventClient.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.meta.wearable.dat.externalsampleapps.cameraaccess.gemini.GeminiConfig
import java.util.UUID
import java.util.concurrent.TimeUnit
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject

class OpenClawEventClient {
    companion object {
        private const val TAG = "OpenClawEventClient"
        private const val MAX_RECONNECT_DELAY_MS = 30_000L
    }

    var onNotification: ((String) -> Unit)? = null

    private var webSocket: WebSocket? = null
    private var isConnected = false
    private var shouldReconnect = false
    private var reconnectDelayMs = 2_000L
    private val handler = Handler(Looper.getMainLooper())

    // Pending RPC responses keyed by request ID
    private val pendingResponses = mutableMapOf<String, (JSONObject) -> Unit>()

    // Pending chat.send results keyed by runId — waits for the "chat" event with state="final"
    private val pendingChatResults = mutableMapOf<String, (String?) -> Unit>()

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(10, TimeUnit.SECONDS)
        .build()

    fun connect() {
        if (!GeminiConfig.isOpenClawConfigured) {
            Log.d(TAG, "Not configured, skipping")
            return
        }
        shouldReconnect = true
        reconnectDelayMs = 2_000L
        establishConnection()
    }

    fun disconnect() {
        shouldReconnect = false
        isConnected = false
        handler.removeCallbacksAndMessages(null)
        // Cancel all pending callbacks so they don't fire after session stops
        pendingResponses.clear()
        pendingChatResults.clear()
        webSocket?.close(1000, null)
        webSocket = null
        Log.d(TAG, "Disconnected")
    }

    private fun establishConnection() {
        val host = GeminiConfig.openClawHost
            .replace("http://", "")
            .replace("https://", "")
        val port = GeminiConfig.openClawPort
        val url = "ws://$host:$port"

        Log.d(TAG, "Connecting to $url")

        val request = Request.Builder()
            .url(url)
            .header("Host", "localhost:${GeminiConfig.openClawPort}")
            .build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket opened")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket failure: ${t.message}")
                isConnected = false
                scheduleReconnect()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closing: $code $reason")
                isConnected = false
                scheduleReconnect()
            }
        })
    }

    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            val type = json.optString("type", "")

            when (type) {
                "event" -> handleEvent(json)
                "res" -> {
                    val id = json.optString("id", "")
                    val callback = pendingResponses.remove(id)
                    if (callback != null) {
                        callback(json)
                    } else {
                        // Connect handshake response
                        val ok = json.optBoolean("ok", false)
                        if (ok) {
                            Log.d(TAG, "Connected and authenticated")
                            isConnected = true
                            reconnectDelayMs = 2_000L
                        } else {
                            val error = json.optJSONObject("error")
                            val msg = error?.optString("message", "unknown") ?: "unknown"
                            Log.e(TAG, "Connect failed: $msg")
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Parse error: ${e.message}")
        }
    }

    private fun handleEvent(json: JSONObject) {
        val event = json.optString("event", "")
        val payload = json.optJSONObject("payload") ?: JSONObject()

        when (event) {
            "connect.challenge" -> sendConnectHandshake()
            "heartbeat" -> handleHeartbeatEvent(payload)
            "cron" -> handleCronEvent(payload)
            "chat" -> handleChatEvent(payload)
        }
    }

    private fun handleChatEvent(payload: JSONObject) {
        val state = payload.optString("state", "")
        val runId = payload.optString("runId", "")

        if (state == "final" && runId.isNotEmpty()) {
            val callback = pendingChatResults.remove(runId)
            if (callback != null) {
                // Extract reply text from message.content
                val message = payload.optJSONObject("message")
                val content = message?.opt("content")
                val replyText = when {
                    content is String -> content
                    content is JSONArray -> {
                        val parts = mutableListOf<String>()
                        for (i in 0 until content.length()) {
                            val part = content.optJSONObject(i)
                            if (part?.optString("type") == "text") {
                                parts.add(part.optString("text", ""))
                            }
                        }
                        parts.joinToString("\n").ifEmpty { null }
                    }
                    else -> null
                }
                Log.d(TAG, "chat final for $runId: ${replyText?.take(200)}")
                callback(replyText ?: "Agent completed but returned no text.")
            }
        } else if (state == "error" && runId.isNotEmpty()) {
            val callback = pendingChatResults.remove(runId)
            if (callback != null) {
                val errorMsg = payload.optString("errorMessage", "Agent error")
                Log.e(TAG, "chat error for $runId: $errorMsg")
                callback(null)
            }
        }
    }

    private fun sendConnectHandshake() {
        val connectMsg = JSONObject().apply {
            put("type", "req")
            put("id", UUID.randomUUID().toString())
            put("method", "connect")
            put("params", JSONObject().apply {
                put("minProtocol", 3)
                put("maxProtocol", 3)
                put("client", JSONObject().apply {
                    put("id", "gateway-client")
                    put("displayName", "VisionClaw Glass")
                    put("version", "1.0")
                    put("platform", "android")
                    put("mode", "backend")
                })
                put("auth", JSONObject().apply {
                    put("token", GeminiConfig.openClawGatewayToken)
                })
                put("scopes", JSONArray().apply {
                    put("operator.admin")
                })
            })
        }
        webSocket?.send(connectMsg.toString())
    }

    private fun handleHeartbeatEvent(payload: JSONObject) {
        val status = payload.optString("status", "")
        if (status != "sent") return

        val preview = payload.optString("preview", "")
        if (preview.isEmpty()) return

        val silent = payload.optBoolean("silent", false)
        if (silent) return

        Log.d(TAG, "Heartbeat notification: ${preview.take(100)}")
        onNotification?.invoke("[Notification from your assistant] $preview")
    }

    private fun handleCronEvent(payload: JSONObject) {
        val action = payload.optString("action", "")
        if (action != "finished") return

        val summary = payload.optString("summary", "").ifEmpty {
            payload.optString("result", "")
        }
        if (summary.isEmpty()) return

        Log.d(TAG, "Cron notification: ${summary.take(100)}")
        onNotification?.invoke("[Scheduled update] $summary")
    }

    /**
     * Send a chat message with optional image attachment via WebSocket chat.send RPC.
     * This is the only way to reliably pass images to the OpenClaw agent.
     * Returns the agent's reply text, or null on failure.
     */
    fun sendChatMessage(
        sessionKey: String,
        message: String,
        imageBase64: String? = null,
        imageMimeType: String = "image/jpeg",
        onResult: (String?) -> Unit
    ) {
        if (!isConnected || webSocket == null) {
            Log.e(TAG, "Cannot send chat.send: not connected")
            onResult(null)
            return
        }

        val reqId = UUID.randomUUID().toString()

        val params = JSONObject().apply {
            put("sessionKey", sessionKey)
            put("message", message)
            put("idempotencyKey", reqId)
            if (imageBase64 != null) {
                put("attachments", JSONArray().put(JSONObject().apply {
                    put("mimeType", imageMimeType)
                    put("fileName", "camera_frame.jpg")
                    put("content", imageBase64)
                }))
            }
        }

        val request = JSONObject().apply {
            put("type", "req")
            put("id", reqId)
            put("method", "chat.send")
            put("params", params)
        }

        // Register callback for RPC ack — then wait for the actual chat event
        pendingResponses[reqId] = { response ->
            val ok = response.optBoolean("ok", false)
            if (ok) {
                // RPC accepted — now wait for the "chat" event with state="final"
                Log.d(TAG, "chat.send accepted, waiting for agent reply (runId=$reqId)")
                pendingChatResults[reqId] = onResult
            } else {
                val error = response.optJSONObject("error")
                val msg = error?.optString("message", "unknown") ?: "unknown"
                Log.e(TAG, "chat.send rejected: $msg")
                onResult(null)
            }
        }

        val sent = webSocket?.send(request.toString()) ?: false
        if (!sent) {
            pendingResponses.remove(reqId)
            Log.e(TAG, "Failed to send chat.send WebSocket message")
            onResult(null)
        } else {
            Log.d(TAG, "chat.send sent (id=$reqId, hasImage=${imageBase64 != null})")
        }
    }

    private fun scheduleReconnect() {
        if (!shouldReconnect) return
        Log.d(TAG, "Reconnecting in ${reconnectDelayMs}ms")
        handler.postDelayed({
            if (shouldReconnect) {
                reconnectDelayMs = (reconnectDelayMs * 2).coerceAtMost(MAX_RECONNECT_DELAY_MS)
                establishConnection()
            }
        }, reconnectDelayMs)
    }
}
